runtime/src/accounts_db.rs

```diff
@@ -2234,10 +2246,32 @@ impl AccountsDb {
             // `lock` released here
         };
 
+        let result = self.read_only_accounts_cache.load(pubkey, slot);
+        if let Some(account) = result {
+            return Some((account, slot));
+        }
+
         //TODO: thread this as a ref
-        self.get_account_accessor_from_cache_or_storage(slot, pubkey, store_id, offset)
+        let mut from_cache = false;
+        let loaded_account = self
+            .get_account_accessor_from_cache_or_storage(slot, pubkey, store_id, offset)
             .get_loaded_account()
-            .map(|loaded_account| (loaded_account.account(), slot))
+            .map(|loaded_account| {
+                from_cache = loaded_account.from_cache();
+                (loaded_account.account(), slot)
+            });
+
+        if self.caching_enabled && !from_cache {
```

Comment:
How about a comment about why the `!from_cache` is sufficient to make sure we aren't loading stale data from the `read-only` cache if there is more up-to-date state in the write cache. Something like:

`We show this store into the read-only cache for account `A` and future loads of `A` from the read-only cache are safe/reflect `A`'s latest state on this fork. 

This safety holds if during replay of slot `S`, we show we only read `A` from the write cache after it's been updated in slot `S`. 

Assume for contradiction this is not true, and we read `A` from the read-only cache *after* it had been updated in `S`. This means an entry `(S, A)` was added to the read-only cache after `A` had been updated in `S`.
 
Now when `(S, A)` was being added to the read-only cache, it must have been true that  `from_cache == false`, which means `(S', A)` does not exist in the write cache yet.  

However, by the assumption for contradiction above ,  `A` has already been updated in `S` which means `(S, A)` must exist in the write cache, which is a contradiction.`